In [7]:
import pandas as pd
pd.set_option('display.max_columns', None)

In [8]:
#Cargar los datos
dflights= pd.read_csv('DelayedFlights.csv')

In [9]:
#Data Frame inicial: 
dflights.shape

(1936758, 30)

In [10]:
dflights

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0         2211.0        2225            WN        335  N712SW   
1         1002.0        1000            WN       3231  N772SW   
2          804.0         750            WN        448  N428WN   
3         1959.0        1925            WN       3920  N464WN   
4         2121.0        2110            WN        378  N726SW   
...          ...         ...           ...        ...     ...   
1936753   1617.0        1552            DL       1621  N938DL   
1936754    904.0         749            DL       1631  N3743H   
1936755   1149.0        1010            DL       1631  N909DA   
1936756   1446.0        1437            DL       1639  N646DL   
1936757   1413.0        1418            DL       1641  N908DL   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
0                    128.0           150.0    116.0     -14.0       8.0   
1                    128.0           145.0    113.0       2.0      19.0   
2                     96.0            90.0     76.0      14.0       8.0   
3                     90.0            90.0     77.0      34.0      34.0   
4                    101.0           115.0     87.0      11.0      25.0   
...                    ...             ...      ...       ...       ...   
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   
1936756              115.0           117.0     89.0       9.0      11.0   
1936757              123.0           135.0    104.0      -5.0       7.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  \
0          IAD  TPA       810     4.0      8.0          0                N   
1          IAD  TPA       810     5.0     10.0          0                N   
2          IND  BWI       515     3.0     17.0          0                N   
3          IND  BWI       515     3.0     10.0          0                N   
4          IND  JAX       688     4.0     10.0          0                N   
...        ...  ...       ...     ...      ...        ...              ...   
1936753    MSP  ATL       906     9.0     18.0          0                N   
1936754    RIC  ATL       481    15.0     34.0          0                N   
1936755    ATL  IAH       689     8.0     32.0          0                N   
1936756    IAD  ATL       533    13.0     13.0          0                N   
1936757    SAT  ATL       874     8.0     11.0          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0               0           NaN           NaN       NaN            NaN   
1               0           NaN           NaN       NaN            NaN   
2               0           NaN           NaN       NaN            NaN   
3               0           2.0           0.0       0.0            0.0   
4               0           NaN           NaN       NaN    

In [11]:
#Reducción del dataset a 200.000 filas
dflights = dflights.sample(n = 200000)

### Descripción de variables y justificación por que se eliminan o mantienen

>**'Unnamed: 0'**,		¿Por qué se elimina?	Tiene la forma de un correlativo de los registros y algunos números no llevan un orden especifico. Al trabajar con pandas de igual forma existe un index, así que se duplica la columna, la ventaja del index de pandas es que mantiene un orden.

> **Year**,	Año, en este caso todos los datos son del año 2008.	¿Por qué se elimina?	Se unifica en una sola columna (Año - mes - día )

>**Month**,	Meses en numeros (1 - 12).	¿Por qué se elimina?	Se unifica en una sola columna (Año - mes - día )

>**DayofMonth'**,	Dia del mes (1-31).	¿Por qué se elimina?	Se unifica en una sola columna (Año - mes - día )

>**DayOfWeek'**,	Dia de la semana (1 - 7 ).	¿Por qué se mantiene?	Puede ser interesante el análisis de los días con más retrasos por aerolínea, aeropuerto.

>**DepTime'**,	Hora de despegue del avion real. Hora en formato militar.	¿Por qué se mantiene?	Se puede analizar cuales son los horarios en los que más hay retrasos.

>**CRSDepTime**,	Hora de despegue del avion planificada o esperada. Hora en formato militar.	¿Por qué se mantiene?	Se mantiene como referencia y futuros análisis. Por ejemplo se podría decir conclusiones como: "El vuelo de las 11pm generalmente tiene retrasos".

>**ArrTime'**,	Hora de aterrizaje del avion real. Hora en formato militar.	¿Por qué se mantiene?	Se puede analizar cuales son los horarios en los que más hay retrasos al momento de arribar. 

>**CRSArrTime**',	Hora de aterrizaje del avion planificada o esperada. Hora en formato militar.	¿Por qué se mantiene?	Se mantiene como referencia y futuros análisis. Por ejemplo se podría decir conclusiones como: "El vuelo que arriba a las 3 am generalmente tiene retrasos".

>**UniqueCarrier'**,	Código unico de la aerolínea.	¿Por qué se mantiene?	Importante para identificar las aerolíneas con mas retrasos.

>**FlightNum'**,	Número de vuelo.	¿Por qué se mantiene?	Se puede identificar que vuelos son los que tienen más retrasos.

>**TailNum'**,	Es el identificador único de los aviones. Cada avión tiene un código generalmente entre 2-6 caracteres. Aunque el nombre indica que debería estar impreso en la "cola" del avión no siempre es así, puede ser impreso en otras partes. Con este identificador puedes encontrar información como el mantenimiento que ha tenido, los lugares que ha viajado, etc.	¿Por qué se mantiene?	Se puede identificar que aviones causan más retrasos.

>**ActualElapsedTime'**,	Tiempo real del avión para despegar y llegar al destino. (minutos).	¿Por qué se mantiene?	Resta entre DepTime y ArrTime. Se puede analizar en que vuelos (corta duración o larga) son más propensas los retrasos.

>**CRSElapsedTime'**,	Tiempo planificado/esperado del avión para despegar y llegar al destino. (minutos).	¿Por qué se mantiene?	Resta entre CRSDepTime y CRSArrTime. Sirve como referencia, por ejemplo, se pueden concluir cosas como "en los vuelos planificados de más de 4 horas tienen a existir más retrasos".

>**AirTime'**,	Tiempo que pasa el avión en el aire. (minutos).	¿Por qué se mantiene?	Es necesario para hacer calculos de velocidad media si se combina con la distancia entre cada aeropuerto.

>**ArrDelay**',	Minutos de retraso al momento de aterrizar.	¿Por qué se mantiene?	Es la resta del tiempo de salida y llegada del avión. En este dato se basa el análisis.

>**DepDelay'**,	Minutos de retraso en el despegue..	¿Por qué se mantiene?	Es la resta entre DepTime y CRSDepTime. Columna útil para los calculos y analizar en que aeropuertos se producen más los retrasos en el despegue.

>**Origin'**,	Código del aeropuerto de origen (IATA= International Air Transport Association).	¿Por qué se mantiene?	Se puede analizar en que aeropuertos se originan muchos retrasos en la salida.

>**>Dest'**,	Código del aeropuerto de destino (IATA= International Air Transport Association).	¿Por qué se mantiene?	Se puede analizar en que aeropuertos se originan muchos retrasos en la llegada.
Dest',	Código del aeropuerto de destino (IATA= International Air Transport Association).	¿Por qué se mantiene?	Se puede analizar en que aeropuertos se originan muchos retrasos en la llegada.

>**Distance'**,	Distancia entre el origen y el destino en millas.	¿Por qué se mantiene?	Es necesario para hacer calculos de velocidad media si se combina con el tiempo en el aire del avión.

>**TaxiIn'**,	Tiempo de entre el aterrizaje del avión y cuando se parquea "activando" los frenos.	¿Por qué se mantiene?	Interesante para analizar el tiempo antes del despegue.

>**TaxiOut'**,	Teimpo desde que "desactiva" los frenos desde su posición de parque y cuando despega el avión.	¿Por qué se mantiene?	Interesante para analizar el tiempo que se tarda el avión despues del aterrizaje. Por ejemplo, puede ser que el avión arribó en tiempo pero no hay suficientes "gates" para aparcar.

>**Cancelled**',	Indica si el vuelo ha sido cancelado o no. 0 = No, 1=Si.	¿Por qué se mantiene?	Se puede analizar que aerolínea cancela más vuelos.

>**CancellationCode'**,	Código que indica la razón de la cancelación del vuelo (A=aerolínea, B=clima, c=NAS, D=seguridad). **NAS: Retrasos y cancelaciones atribuibles al sistema de aviación nacional que se refieren a un amplio conjunto de condiciones, tales como condiciones climáticas no extremas, operaciones aeroportuarias, volumen de tráfico pesado y control del tráfico aéreo.	¿Por qué se mantiene?	Se puede analizar por que se cancelan la mayor cantidad de vuelos en el año

>**Diverted'**,	Indica si el vuelo ha sido desviado de su trayectoria original. 1 = SI, 0 = NO.	¿Por qué se mantiene?	Se puede analizar en que rutas generalmente existen más desviaciones. 

>**CarrierDelay**	Retrasos debido a la aerolínea en minutos.	¿Por qué se mantiene?	Se puede analizar el % del tiempo de retraso asociado a la variable.

>**LateAircraftDelay**	Retrasos debido a que el mismo avión llego tarde y en consecuencia el siguiente vuelo tiene retraso.	¿Por qué se mantiene?	Se puede analizar el % del tiempo de retraso asociado a la variable.

>**WeatherDelay**	Retrasos debido al clima en minutos.	¿Por qué se mantiene?	Se puede analizar el % del tiempo de retraso asociado a la variable.

>**NASDelay**	Retrasos debido a la NAS en minutos.	¿Por qué se mantiene?	Se puede analizar el % del tiempo de retraso asociado a la variable.

>**SecurityDelay**	Retrsos debido a las seguridad en minutos.	¿Por qué se mantiene?	Se puede analizar el % del tiempo de retraso asociado a la variable.

In [13]:
#Reducir 3 columnas a 1 (Year-Month-DayofMonth)
dflights['DepDate'] = pd.to_datetime(dflights.Year*10000+dflights.Month*100+dflights.DayofMonth,format='%Y%m%d')

In [14]:
#Eliminar columnas
dflights = dflights[[#'Unnamed: 0', 
     #'Year',
    #'Month', 
    #'DayofMonth',
    'DayOfWeek', 'DepTime',
    'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
    'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
    'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
    'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
    'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay','DepDate']].copy()

In [15]:
dflights.head()

DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  \
1676530          5    855.0         800   1106.0        1036            US   
224897           2   1512.0        1430   1729.0        1535            XE   
553857           3   1903.0        1810   2206.0        2100            B6   
412934           5   1441.0        1355   1541.0        1458            XE   
187831           2   1946.0        1915   2213.0        2150            WN   

         FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  \
1676530        987  N205UW              131.0           156.0    118.0   
224897        2975  N17524              137.0            65.0     55.0   
553857        1419  N274JB              183.0           170.0    161.0   
412934         120  N14186               60.0            63.0     49.0   
187831        3135  N276WN              147.0           155.0    135.0   

         ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled  \
1676530      30.0      55.0    PHL  MCO       861     2.0     11.0          0   
224897      114.0      42.0    DAL  IAH       217     9.0     73.0          0   
553857       66.0      53.0    BOS  JAX      1011     4.0     18.0          0   
412934       43.0      46.0    MRY  ONT       296     5.0      6.0          0   
187831       23.0      31.0    PBI  ISP      1052     4.0      8.0          0   

        CancellationCode  Diverted  CarrierDelay  WeatherDelay  NASDelay  \
1676530                N         0          30.0           0.0       0.0   
224897                 N         0           0.0           0.0      72.0   
553857                 N         0           0.0           0.0      13.0   
412934                 N         0           5.0           0.0       0.0   
187831                 N         0           1.0           0.0       0.0   

         SecurityDelay  LateAircraftDelay    DepDate  
1676530            0.0                0.0 2008-11-28  
224897             0.0               42.0 2008-02-05  
553857             0.0               53.0 2008-03-19  
412934             0.0               38.0 2008-03-14  
187831             0.0               22.0 2008-02-05

In [16]:
# Shape del dataset con el que realizaremos el ejercicios
dflights.shape

(200000, 27)

In [17]:
#Visualizamos los diferentes tipos de datos 
dflights.dtypes

DayOfWeek                     int64
DepTime                     float64
CRSDepTime                    int64
ArrTime                     float64
CRSArrTime                    int64
UniqueCarrier                object
FlightNum                     int64
TailNum                      object
ActualElapsedTime           float64
CRSElapsedTime              float64
AirTime                     float64
ArrDelay                    float64
DepDelay                    float64
Origin                       object
Dest                         object
Distance                      int64
TaxiIn                      float64
TaxiOut                     float64
Cancelled                     int64
CancellationCode             object
Diverted                      int64
CarrierDelay                float64
WeatherDelay                float64
NASDelay                    float64
SecurityDelay               float64
LateAircraftDelay           float64
DepDate              datetime64[ns]
dtype: object

## Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.


In [18]:
dflights_estadistico= dflights [[ #'DayOfWeek', 
                        'DepTime', 
                        'CRSDepTime', 
                        'ArrTime', 
                        'CRSArrTime',
                        'UniqueCarrier', 
                        #'FlightNum', 
                        'TailNum',
                        'ActualElapsedTime',
                        'CRSElapsedTime',
                        'AirTime', 
                        'ArrDelay',
                        'DepDelay', 
                        #'Origin',
                        #'Dest',
                        'Distance',
                        'TaxiIn',
                        'TaxiOut',
                        'Cancelled',   
                        #'CancellationCode',
                        'Diverted',
                        'CarrierDelay', 
                        'WeatherDelay', 
                        'NASDelay', 
                        'SecurityDelay',
                        'LateAircraftDelay','DepDate']].copy()

In [19]:
dflights_estadistico.describe()

DepTime     CRSDepTime        ArrTime     CRSArrTime  \
count  200000.000000  200000.000000  199247.000000  200000.000000   
mean     1518.988840    1468.000105    1608.960311    1633.712900   
std       451.517053     425.645831     549.633816     465.926395   
min         1.000000       1.000000       1.000000       1.000000   
25%      1203.000000    1135.000000    1314.000000    1325.000000   
50%      1546.000000    1510.000000    1715.000000    1707.000000   
75%      1900.000000    1816.000000    2030.000000    2015.000000   
max      2400.000000    2359.000000    2400.000000    2400.000000   

       ActualElapsedTime  CRSElapsedTime        AirTime       ArrDelay  \
count      199124.000000   199979.000000  199124.000000  199124.000000   
mean          132.919241      133.911391     107.884389      42.312338   
std            71.676168       70.878588      68.181639      57.155150   
min            14.000000      -25.000000       0.000000     -66.000000   
25%            80.000000       82.000000      58.000000       9.000000   
50%           116.000000      116.000000      90.000000      24.000000   
75%           164.000000      165.000000     137.000000      56.000000   
max           697.000000      660.000000     654.000000    1542.000000   

            DepDelay       Distance         TaxiIn        TaxiOut  \
count  200000.000000  200000.000000  199247.000000  199955.000000   
mean       43.274535     762.369435       6.818150      18.235963   
std        53.822812     570.773713       5.285628      14.444391   
min         6.000000      31.000000       0.000000       0.000000   
25%        12.000000     338.000000       4.000000      10.000000   
50%        24.000000     606.000000       6.000000      14.000000   
75%        53.000000     992.000000       8.000000      21.000000   
max      1545.000000    4962.000000     186.000000     335.000000   

           Cancelled       Diverted   CarrierDelay   WeatherDelay  \
count  200000.000000  200000.000000  128994.000000  128994.000000   
mean        0.000315       0.004065      19.463673       3.620044   
std         0.017745       0.063628      44.285529      21.582105   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       2.000000       0.000000   
75%         0.000000       0.000000      21.000000       0.000000   
max         1.000000       1.000000    1542.000000    1352.000000   

            NASDelay  SecurityDelay  LateAircraftDelay  
count  128994.000000  128994.000000      128994.000000  
mean       15.022582       0.092493          25.172372  
std        33.877796       2.109654          41.654960  
min         0.000000       0.000000           0.000000  
25%         0.000000       0.000000           0.000000  
50%         1.000000       0.000000           8.000000  
75%        15.000000       0.000000          33.000000  
max      1289.000000     284.000000        1000.000000

## Análisis estadístico


**DepTime & CRSDepTime**
> Entre el 25% y el 75% esta previsto que despeguen en un horario de 11:35 y 18:16, sin embargo, en la realidad despegan más tarde, entre 12:03 y 19:00 horas.

**ArrTime & CRSArrTime**
> Entre el Q1(25%) y Q3(75%) los vuelos deberían arribar entre las 13:25 y 20:15, en la realidad arriban entre las 13:14 y las 20:30. Si lo comparamos con los horarios de despegues los valores previstos no varian en la misma proporción con los reales. Esto se puede deber a que los retrasos en el despegue en cierta forma estan contemplados y de esta forma se intentan reducir los retrasos en la hora de llegada. *Hay que tomar en cuenta que en ArrTime se tienen 199247 datos y en CRS ArrTime los 200000.
 
 **ActualElapsedTime & CRSElapsedTime**
> Algo interesante a comparar en este caso son las medias de ActualElapsedTime(media=132.91 minutos) y CRSElapsedTime(media=133.91 minutos), lo que indica que en tiempo promedio los vuelos arriban más rápido que lo planificado. Este resultado puede estar sesgado por algunos outliers donde algún vuelo siempre arribe antes de tiempo.

**Airtime**

> El tiempo promedio que una avión esta en el aire en los vuelos del año 2008 es de 108 minutos.

**ArrDelay & **DepDelay****

> En promedio se tienen 42.31 minutos de retraso en los arribos. Entre 9 y 56 minutos son los retrasos que se pueden esperar entre el Q1(25%) y Q3(75%) de los vuelos.

> En promedio se tienen 43.27 minutos de retraso en los despegues.

> Se puede observar que existen más retrasos en el despegue que en el momento del arribo al otro aeropuerto.

**Distance**
> La distancia media de los vuelos es de 762 millas, el mínimo de distancia es de 31 millas y el máximo de 4962 millas. El 75% de los vuelos no sobrepasan las distancia de 992 millas. 

**TaxiIn & TaxiOut**
> El tiempo promedio qu ese tarda el avión desde que sale del "parqueo", va a la pista y  despega (18.24 mins) es mayor que el tiempo que se tarda el avión desde el aterrizaje hasta que esta en el parqueo y activa los frenos(6.81 mins). En las distintas variables analizadas se puede observar que el retraso en el despegue es mucho mayor que en el del aterrizaje.

**Cancelled**

>El 0.0315% fueron cancelados de la muestra de 200.000 vuelos, lo que serían: 63 vuelos. 

**Diverted**

>El 0.4065% fueron desviados de la muestra de 200.000 vuelos, lo que serían: 813 vuelos. 

**Distintos Retrasos**
> **Carrier Delay** El tiempo de retraso medio debido a las aerolíneas es de: 19.43 minutos.

> **Weather Delay** El tiempo de retraso medio debido al clima es de: 3.62 minutos.

> **NAS Delay** El tiempo de retraso medio debido a la NAS(National Airspace System) es de: 15.02 minutos.

> **Security Delay** El tiempo de retraso medio debido a la seguridad  es de: 0.09 minutos.

> **LateAircraft Delay** El tiempo de retraso medio debido al retraso del avión en algún vuelo anterior es de: 25.17  minutos.

El "Late Aircraft delay" es la causa de retrasos que más tiempo afecta los vuelos diaros. Es lo que se considera como el efecto Ripple, consiste por ejemplo, si un avión llega tarde todos los demás tendrán cierto retraso. Es decir, si un avión llegua tarde probablemente afectara a todos los vuelos siguientes.

El segundo más importante son los retrasos que se atribuyen a la aerolínea(19.43 minutos).

### Análizando el ArrDelay con las otras variables. 

Una variable muy importante y crítica es la que indica el retraso en el arribo. Por lo cual, a continuación se comparara con las otras variables para ver en que porcentaje le afecta y esta relacionada. Para eso se realiza un matriz de correlación

In [20]:
dflights_corr= dflights[[
                #'DayOfWeek', 
                #'DepTime', 
                #'CRSDepTime', 
                #'ArrTime', 
                #'CRSArrTime',
                #'UniqueCarrier',
                #'FlightNum', 
                #'TailNum', 
                #'ActualElapsedTime',
                'ArrDelay',
                'CRSElapsedTime', 
                #'AirTime', 
                'DepDelay', 
                #'Origin', 
                #'Dest',
                'Distance', 
                'TaxiIn', 
                'TaxiOut', 
                #'Cancelled', 
                #'CancellationCode',
                #'Diverted', 
                'CarrierDelay', 
                'WeatherDelay', 
                'NASDelay', 
                'SecurityDelay',
                'LateAircraftDelay']].copy()
dflights_corr.corr()

ArrDelay  CRSElapsedTime  DepDelay  Distance    TaxiIn  \
ArrDelay           1.000000       -0.010291  0.953253 -0.024700  0.155997   
CRSElapsedTime    -0.010291        1.000000  0.010986  0.981294  0.100714   
DepDelay           0.953253        0.010986  1.000000 -0.004149  0.073563   
Distance          -0.024700        0.981294 -0.004149  1.000000  0.073145   
TaxiIn             0.155997        0.100714  0.073563  0.073145  1.000000   
TaxiOut            0.287000        0.126798  0.079225  0.072466  0.069037   
CarrierDelay       0.514246        0.028134  0.541095  0.029735 -0.019788   
WeatherDelay       0.268135       -0.017709  0.249490 -0.024512  0.029713   
NASDelay           0.385795        0.052287  0.232504  0.022078  0.229645   
SecurityDelay      0.000567        0.008394  0.000023  0.008696 -0.005862   
LateAircraftDelay  0.470783       -0.008525  0.504204 -0.007176 -0.010462   

                    TaxiOut  CarrierDelay  WeatherDelay  NASDelay  \
ArrDelay           0.287000      0.514246      0.268135  0.385795   
CRSElapsedTime     0.126798      0.028134     -0.017709  0.052287   
DepDelay           0.079225      0.541095      0.249490  0.232504   
Distance           0.072466      0.029735     -0.024512  0.022078   
TaxiIn             0.069037     -0.019788      0.029713  0.229645   
TaxiOut            1.000000     -0.028965      0.074041  0.432766   
CarrierDelay      -0.028965      1.000000     -0.065882 -0.119306   
WeatherDelay       0.074041     -0.065882      1.000000 -0.004023   
NASDelay           0.432766     -0.119306     -0.004023  1.000000   
SecurityDelay     -0.001038     -0.018631     -0.006587 -0.010484   
LateAircraftDelay -0.050552     -0.174985     -0.050249 -0.116635   

                   SecurityDelay  LateAircraftDelay  
ArrDelay                0.000567           0.470783  
CRSElapsedTime          0.008394          -0.008525  
DepDelay                0.000023           0.504204  
Distance                0.008696          -0.007176  
TaxiIn                 -0.005862          -0.010462  
TaxiOut                -0.001038          -0.050552  
CarrierDelay           -0.018631          -0.174985  
WeatherDelay           -0.006587          -0.050249  
NASDelay               -0.010484          -0.116635  
SecurityDelay           1.000000          -0.018065  
LateAircraftDelay      -0.018065           1.000000

### Observaciones de la matriz de correlaciones

> Se puede observar que no existe una correlación entre la distancia y el tiempo de retraso. Es decir, no esta relacionado que un vuelo por ser más largo tiene más probabilidad de en arribar con retraso.

> Lo que era lógico imaginar es que si hay retraso en el despegue también llegara con retraso al destino en la mayoría de ocasiones.

> Un análisis cruzado puede surgir de analizar el Carrier delay vs { ArrDelay (0.51 correlacion) y DepDelay(0.54 correlacion)} , es decir que el Carrier delay afecta más en el retraso asociado al despegue que al del arribo.

>En el caso del WeatherDelay vs { ArrDelay(0.26 correlacion) y DepDelay (0.24 correlacion) }, el caso es contrario afecta más en el arribo que en el despegue. 

>El NAS delay afecta más en los arribos que en el despegue. Por ejemplo, si no hay espacio en pista para que te aterrice le pediran que este en el aire más tiempo.


## Troba quantes dades faltants hi ha per columna.

In [21]:
dflights.isna().sum()

DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                753
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                  2
ActualElapsedTime      876
CRSElapsedTime          21
AirTime                876
ArrDelay               876
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
TaxiIn                 753
TaxiOut                 45
Cancelled                0
CancellationCode         0
Diverted                 0
CarrierDelay         71006
WeatherDelay         71006
NASDelay             71006
SecurityDelay        71006
LateAircraftDelay    71006
DepDate                  0
dtype: int64

## Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...)

In [22]:
#Agregando una columna con la velocidad media en el aire (millas / horas)
dflights['VelMediaVuelo'] = (dflights['Distance'] / ((dflights['AirTime'])/60) )

In [23]:
# Se agrega una columna si arriba tarde o no, 1=Tarde, 0=en tiempo, se considera tarde si arriba después de 15 minutos. 
dflights['EstadoVuelollegada'] = ["Tarde" if x > 15 else "En tiempo" for x in dflights['ArrDelay']]

In [24]:
#Agrego una columna con los nombres de las aerolíneas
airlines= pd.read_csv('https://raw.githubusercontent.com/beanumber/airlines/master/data-raw/airlines.csv')
airlines = airlines.rename(columns ={"Code": "UniqueCarrier"})
dflights = dflights.merge(airlines, how='left', on='UniqueCarrier')

In [25]:
#Número total de vuelos por aerolínea
dflights['Description'].value_counts()

Southwest Airlines Co.          39063
American Airlines Inc.          19929
Envoy Air                       14677
United Air Lines Inc.           14561
SkyWest Airlines Inc.           13688
Delta Air Lines Inc.            11624
ExpressJet Airlines Inc. (1)    10801
Continental Air Lines Inc.      10186
US Airways Inc.                 10171
ExpressJet Airlines Inc.         8482
Northwest Airlines Inc.          8208
AirTran Airways Corporation      7328
Mesa Airlines Inc.               6986
JetBlue Airways                  5789
Comair Inc.                      5442
Endeavor Air Inc.                5349
Alaska Airlines Inc.             3944
Frontier Airlines Inc.           2950
Hawaiian Airlines Inc.            738
Name: Description, dtype: int64

In [26]:
#Cantidad de vuelos con llegada tarde
dflights_tarde= dflights.query("EstadoVuelollegada=='Tarde'")
dflights_tarde["Description"].value_counts().sort_values( ascending=False)


Southwest Airlines Co.          20309
American Airlines Inc.          13441
Envoy Air                        9825
United Air Lines Inc.            9591
SkyWest Airlines Inc.            8884
ExpressJet Airlines Inc. (1)     7343
Delta Air Lines Inc.             7186
US Airways Inc.                  5946
Continental Air Lines Inc.       5914
ExpressJet Airlines Inc.         5789
Northwest Airlines Inc.          5592
Mesa Airlines Inc.               5211
AirTran Airways Corporation      4679
Comair Inc.                      3940
JetBlue Airways                  3940
Endeavor Air Inc.                3603
Alaska Airlines Inc.             2410
Frontier Airlines Inc.           1578
Hawaiian Airlines Inc.            414
Name: Description, dtype: int64

In [27]:
#Quins són els vols més llargs?
dflights[['FlightNum','Origin','Dest','Distance',]].sort_values('Distance',ascending=False).drop_duplicates()

FlightNum Origin Dest  Distance
81062          15    EWR  HNL      4962
129302         14    HNL  EWR      4962
191429       1282    HNL  ATL      4502
46894         850    HNL  ATL      4502
136722       1560    HNL  ATL      4502
...           ...    ...  ...       ...
107220         77    GST  JNU        41
116357       5463    LAX  SNA        36
135482       5471    LAX  SNA        36
60193          64    PSG  WRG        31
17678          65    WRG  PSG        31

[56423 rows x 4 columns]

In [28]:
#Quins són els més endarrerits??
dflights[['FlightNum','Origin','Dest','Distance','ArrDelay']].sort_values('ArrDelay',ascending=False).drop_duplicates().head(15)

FlightNum Origin Dest  Distance  ArrDelay
32058        1472    MOT  MSP       449    1542.0
19113        4333    DSM  ORD       299    1417.0
33014        2093    LGA  MIA      1097    1370.0
126018       1790    LAS  STL      1372    1225.0
180815       1077    TPA  IND       838    1203.0
51180        4801    CVG  LGA       585    1195.0
60485        1965    SDF  DFW       733    1193.0
12984        1226    DEN  DTW      1123    1186.0
4628          175    LAS  MEM      1416    1170.0
53471        3744    LBB  DFW       282    1148.0
16463         243    FLL  DTW      1127    1146.0
166805       1280    DEN  ORD       888    1143.0
97639         286    LIH  LAX      2615    1113.0
86292         268    SEA  JFK      2421    1098.0
57917         170    SEA  MSP      1399    1097.0

In [29]:
#Es interesante calcular los días donde hay mas retrasos
# El día domingo es donde se pueden observar más retrasos.
dflights.groupby('DayOfWeek')['ArrDelay'].mean().sort_values(ascending=False)

DayOfWeek
7    44.540014
2    44.351093
5    43.819850
1    41.581580
4    41.300677
3    40.078368
6    39.733812
Name: ArrDelay, dtype: float64

In [30]:
#Retrasos en día domingo
dflights.groupby('DayOfWeek')['ArrDelay'].mean().sort_values(ascending=False)


DayOfWeek
7    44.540014
2    44.351093
5    43.819850
1    41.581580
4    41.300677
3    40.078368
6    39.733812
Name: ArrDelay, dtype: float64

In [42]:
#Aerolíneas con más retrasos en promedio, se elige CarrierDelay 
dflights.groupby('Description')['CarrierDelay'].mean().sort_values(ascending=False)

Description
Mesa Airlines Inc.              37.230683
Hawaiian Airlines Inc.          37.112644
ExpressJet Airlines Inc.        29.940689
Comair Inc.                     28.193221
Northwest Airlines Inc.         27.666201
Endeavor Air Inc.               25.448753
American Airlines Inc.          21.832629
Delta Air Lines Inc.            19.914003
Alaska Airlines Inc.            19.079824
US Airways Inc.                 18.899984
SkyWest Airlines Inc.           18.819186
Envoy Air                       17.804822
JetBlue Airways                 17.793766
United Air Lines Inc.           17.790916
ExpressJet Airlines Inc. (1)    17.277386
Continental Air Lines Inc.      16.905520
Frontier Airlines Inc.          16.855335
Southwest Airlines Co.          11.377170
AirTran Airways Corporation      9.211367
Name: CarrierDelay, dtype: float64

## Exercici 3
Exporta el dataset net i amb les noves columnes a Excel.

In [44]:
dflights.to_excel('Vuelos_200000.xlsx', index=False)